Preparation for CRAN submission

marohrdanz opened the pull request:
This pull request contains changes for initial submission to CRAN.

Here are some key changes:

- Changed the location for temporary files to the R session's tempdir() (9cb8ead20344bd9a72f02917c2621ed310b6cfce)
- Changes to C code for so that examples run and tests pass on Windows (b3293505ef2cef09b201ef553527d2ff2d226e14, 44eaf1925bae3703f7e6a7b1afd33bcd022bc3ec).
- Changes to C code to eliminate C compiler warning on Windows (f445393f5120c78b8107691dab1e7efafc6d252b, 3dc100ec6ce3c1c9754c795bf2384c3d1ddbc17a).
- Changes to examples to eliminate warning about unused connections (f1bb5608d3a06c38aae90dd8831707c451eb0575).
- Minor changes to eliminate a NOTE and WARNING from `R CMD check --as-cran` (bd35ffcab23e9fba080e48b530fa3397a267cdb8, 20d4d2470d285c0826ebbba629a2000a157acdfb)
- Change of URL in DESCRIPTION (efe69258b6d3465dc3f6cf208972489f49e31da4).
- Change of maintainer to @marohrdanz (Let me know if I should change it back to @bmbroom for actual CRAN submission). (91143357b206f70dd1c7ce9d5a85a7bbf0687a4e).
- Update version number to 1.0.0 (8612e93682e0d9692b1bf1c1b09e796a83fa93a5).

With the changes in this pull request, running `R CMD check --as-cran` gives no ERRORS or WARNINGS, and only one NOTE about this being a new submission. The same is true when using the [Winbuilder](https://win-builder.r-project.org/). The package also builds successfully on [R-hub builder](https://builder.r-hub.io/). These builders are specified in the [Checklist for CRAN submissions](https://cran.r-project.org/web/packages/submission_checklist.html)

marohrdanz committed:
Add *.dll to .gitignore

marohrdanz committed:
Patch for examples & tests to work on windows

In testing for CRAN, we discovered the examples and tests fail under
windows. This commit addresses that (so the examples run and the tests
pass). Also included a few formatting changes to avoid a compiler
warning that appears on windows regarding misleading indentation in a
few if statements.

marohrdanz committed:
Use write.table instead of writeLines in examples

Making this change to avoid warning messages regarding unused
connections. (I couldn't quickly find a way to close the connection to
the datafile with writeLines.)

marohrdanz committed:
List @marohrdanz as maintainer

Making this change (at least temporarily) for checking the build on
CRAN's windows check service (https://win-builder.r-project.org). Their
instructions point out that only the listed maintainer will receive
email notifications regarding the builds.

marohrdanz committed:
Update version number: 0.0.14 -> 1.0.0

marohrdanz committed:
Remove 'Date' field from DESCRIPTION

If this date is more than 1 month old, it results in a NOTE from `R CMD
check --as-cran`. The O'Reilly book on R packages discourages the use of
the Date field, as it is easy to forget to update.
Ref: https://r-pkgs.org/description.html

marohrdanz committed:
Make example lines less than 100 characters

If an example line is more than 100 characters, a WARNING is generated
by R CMD check --as-cran. This warning states that such lines will be
truncated in the PDF.

I did try simply adding newline characters in these examples, but the
warnings persisted.

marohrdanz committed:
Change DESCRIPTION URL to GitHub repo

This is the URL that will appear on the first page of CRAN's
documentation for the package. I'm guessing @bmbroom doesn't want it to
point to his faculty profile. For reference, the RPPA SPACE package also
lists the GitHub repo as its URL.

marohrdanz committed:
Include strndup definition for windows compiler

Building on Windows gives the following warning about the use of memcpy:

warning: passing argument 1 of 'memcpy' discards 'const' qualifier from
         pointer target type [-Wdiscarded-qualifiers]
  210 |             memcpy (dht->slot[idx].str, str, len);
      |                     ~~~~~~~~~~~~~~^~~~

It looks like the block of code containing this memcpy call was added
to account for strndup not being in <string.h> on Windows systems.

Because of the compiler warning, I removed this block of code and
replaced it with the definition of strdnup from this SO:
  https://stackoverflow.com/questions/46013382/
      c-strndup-implicit-declaration

ChrisWakefield reviewed (changes requested):
I raised some questions to Mary.
A broader question is whether we want the windows package to only be able to read windows files and linux its files (referring to line-endings).

jmelott commented:
Hi Chris,

The current settings would be the most common and people are used to having to deal with line ending changes already if they’re moving files between environments.
I could see the line ending as-is being the default, but I could see the use for allowing someone to specify which type of line ending they want to use.

I could also see a use case regarding which separator character (only tabs now) and/or which character they want to use for decimal point characters as we already had a request in the past that we be able to handle using commas instead of decimal points.  The majority of European countries use the decimal comma, the UK being one of the notable exceptions.

Although these changes might be the best options in the long run, none of them are necessary for a first submittal to CRAN and could always be added as enhancements later.  Having the entire software NG-CHM suite in CRAN would help visibility and adoption of the software and might help secure funding to pay for such enhancements.

James M. Melott


From: Chris Wakefield ***@***.***>
Sent: Thursday, September 7, 2023 5:38 PM
To: MD-Anderson-Bioinformatics/tsvio ***@***.***>
Cc: Melott,James M ***@***.***>; Review requested ***@***.***>
Subject: [EXTERNAL] Re: [MD-Anderson-Bioinformatics/tsvio] Preparation for CRAN submission (PR #2)

THIS EMAIL IS A PHISHING RISK


Do you trust the sender?
The email address is: ***@***.***
While this email has passed our filters, we need you to review with caution before taking any action.
If the email looks at all suspicious, click the Report a Phish button.



@ChrisWakefield requested changes on this pull request.

I raised some questions to Mary.
A broader question is whether we want the windows package to only be able to read windows files and linux its files (referring to line-endings).

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/MD-Anderson-Bioinformatics/tsvio/pull/2*pullrequestreview-1616328976__;Iw!!PfbeBCCAmug!j2lEA6R12CFZ97NrgpL8gdK0MBLhkabtv01TNDOzyepMfBl6sr75l3X97rReVW5rS-e0eAvAysDvPDqXFrK0y5EaN30$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/ABYQRLGH7MBRVFLGUKY7ANLXZJEFZANCNFSM6AAAAAA4PMPOVU__;!!PfbeBCCAmug!j2lEA6R12CFZ97NrgpL8gdK0MBLhkabtv01TNDOzyepMfBl6sr75l3X97rReVW5rS-e0eAvAysDvPDqXFrK041kRhxk$>.
You are receiving this because your review was requested.Message ID: ***@***.***>

The information contained in this e-mail message may be privileged, confidential, and/or protected from disclosure. This e-mail message may contain protected health information (PHI); dissemination of PHI should comply with applicable federal and state laws. If you are not the intended recipient, or an authorized representative of the intended recipient, any further review, disclosure, use, dissemination, distribution, or copying of this message or any attachment (or the information contained therein) is strictly prohibited. If you think that you have received this e-mail message in error, please notify the sender by return e-mail and delete all references to it and its contents from your systems.

marohrdanz committed:
Cleanup strndup implementation for windows

Added some code comments. Also fixed a bug pointed out by
@ChrisWakefield in the malloc call: n1 is the correct choice because the
code determines n1 explicitly to be the length of s.

marohrdanz committed:
Refactor handling of '/r' in get_tsv_line_buffer

In b329350 I made changes to test for '\r' in this while loop. Those
changes resulted in a need for #ifdef _WIN32 block. Our plan is to move
away from needed any such blocks in the actual code. So doing this
refactor (suggested by @ChrisWakefield) to remove this _WIN32 instance.

marohrdanz committed:
Remove _WIN around unlink call

Testing on windows shows the unlink function is available. The Microsoft
documentation states the function is deprecated, and suggests either
using `_unlink()`, or turning off warnings. Before submission I'll check
if the CRAN windows compiler gives any warnings (my compiler on windows
does not).

Microsoft reference documentation:
https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference
  /unlink?view=msvc-170

marohrdanz committed:
Write temporary files in tempdir()

CRAN's policies state packages should not write anywhere "on the file
system apart from the R sessions's temporary directory". This commit
implements that here (for both Windows and linux).

Reference:
https://cran.r-project.org/web/packages/policies.html

marohrdanz committed:
Break at line ending in scan_header_line

When determining the column headings in scan_header_line, stop examining
when we get to a line ending. This change is to account for Windows line
endings.

marohrdanz committed:
Update README to use MD-Anderson-Bioinformatics

The old instructions still referred to the bmbroom repository.

jmelott reviewed (commented):
Shouldn't it have instructions on how to install the latest from CRAN as well as that would be the normal way to install it and this is the secondary method if the CRAN version has issues or someone wants to get the latest?

marohrdanz commented:
@jmelott , yes...but I didn't want CRAN install instructions in the default GitHub branch until the package has actually been accepted by CRAN.

jmelott commented:
Oh wait, that particular file is only in the the Repo, not part of the CRAN package.  My mistake.  That should be OK.
